### PR TITLE
refactor: improve build

### PR DESCRIPTION
**Pull request by f0rmiga**

### CODE REVIEW

File: image/Dockerfile [hunk 1]
```diff
@@ -12,23 +12,41 @@
...
-FROM debian:stretch
+# --------------------------------------------------------------------------------------------------
+# The building stage.
+FROM golang:1.14.2-stretch AS builder
 
-RUN apt-get update
-RUN apt-get install -y curl ca-certificates
+WORKDIR /build/minibroker
+COPY ./go.mod ./go.mod
+COPY ./go.sum ./go.sum
```
Comment (mook-as) [inline image/Dockerfile:21]:
```suggestion
# Copy the go.mod over so docker can cache the module downloads if possible.
COPY go.mod go.mod ./
```

File: image/Dockerfile [hunk 2]
```diff
@@ -12,23 +12,41 @@
...
+
+COPY ./ ./
+RUN make build-linux
+
+# --------------------------------------------------------------------------------------------------
+# The downloading stage.
+FROM alpine:latest AS downloader
 
 ENV HELM_VERSION="v2.13.1"
 ENV FILENAME="helm-${HELM_VERSION}-linux-amd64.tar.gz"
```
Comment (mook-as) [inline image/Dockerfile:32]: Should these turn into build args instead?

File: image/Dockerfile [hunk 3]
```diff
@@ -12,23 +12,41 @@
...
-RUN curl -L http://storage.googleapis.com/kubernetes-helm/${FILENAME} -o /tmp/${FILENAME} && \
-    tar -zxvf /tmp/${FILENAME} -C /tmp && \
-    mv /tmp/linux-amd64/helm /usr/local/bin/
+RUN wget -O /tmp/${FILENAME} http://storage.googleapis.com/kubernetes-helm/${FILENAME}
+RUN tar -zxvf /tmp/${FILENAME} -C /tmp
 
-COPY minibroker /usr/local/bin/
+# --------------------------------------------------------------------------------------------------
+# The running stage.
+FROM debian:stretch
```
Comment (mook-as) [inline image/Dockerfile:39]: Consider updating to stable (possibly in a separate PR):
```suggestion
FROM debian:stable
```

File: image/Dockerfile [hunk 4]
```diff
@@ -12,23 +12,41 @@
...
+
+RUN apt-get update
+RUN apt-get install -y ca-certificates
```
Comment (mook-as) [inline image/Dockerfile:42]: We only need the Debian base for the certs right?
I wonder if we can just `ADD` the file from https://curl.haxx.se/docs/caextract.html in the right place… I guess we should explore that in a separate PR.

Either way we should probably:
```suggestion
RUN apt update && apt install -y ca-certificates && apt clean && rm -rf /var/lib/apt/lists/*
```

File: third-party/go.mod [hunk 5]
```diff
@@ -0,0 +1 @@
+/* A file to exclude the third-party directory from the top-level Go modules processing. */
```
[the remark this answers is not in the document]
Comment (f0rmiga) [reply third-party/go.mod:1]: Oh, thanks for catching that. I'll assume the official docs are the correct one here. :sweat_smile:

File: image/Dockerfile (hunk 3 above)
Comment (f0rmiga) [reply image/Dockerfile:39]: I'll do this in a separate PR where I also make the images configurable.

File: image/Dockerfile (hunk 2 above)
Comment (f0rmiga) [reply image/Dockerfile:32]: I'll have a follow-up PR with configurable args, they will include the image bases and this one.

File: image/Dockerfile (hunk 1 above)
Comment (f0rmiga) [reply image/Dockerfile:21]: Could you update the suggestion to `go.{mod,sum}`? I don't want to add the same file twice. :sweat_smile:

File: image/Dockerfile [hunk 6]
```diff
@@ -12,23 +12,41 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-FROM debian:stretch
+# --------------------------------------------------------------------------------------------------
+# The building stage.
+FROM golang:1.14.2-stretch AS builder
```
[the remark this answers is not in the document]
Comment (f0rmiga) [reply image/Dockerfile:17]: Okay, I'll also put this in the next PR.

File: image/Dockerfile (hunk 1 above)
Comment (mook-as) [reply image/Dockerfile:21]: Like this then?
```suggestion
COPY go.mod go.sum ./
```

Comment (f0rmiga) [reply image/Dockerfile:21]: :crossed_fingers:

**Pull request merged**